machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,218 @@
+}
+
+// MachineOSBuildSpec describes user-configurable options as well as information about a build process.
+type MachineOSBuildSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineConfigPoolReference `json:"machineConfigPool"`
+	// buildInputs is where user options for the build live
+	// +kubebuilder:validation:Required
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// currentConfig is the currently running config on the MCP
+	// +optional
+	CurrentConfig RenderedMachineConfigReference `json:"currentConfig"`
+	// desiredConfig is the desired config we want to build an image for. If currentConfig and desiredConfig are not the same, we need to build an image.
+	// +kubebuilder:validation:Required
+	DesiredConfig RenderedMachineConfigReference `json:"desiredConfig"`
+}
+
+// MachineOSBuildStatus describes the state of a build and other helpful information.
+type MachineOSBuildStatus struct {
+	// conditions are state related conditions for the build. Valid types are:
+	// BuildPrepared, Building, BuildFailed, BuildInterrupted, BuildRestarted, and Ready
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// renderedMachineOSImage describes the machineOsImage object created to track image specific information
+	// machineConfig is a reference to the MC that was used to build this image.
+	// +optional
+	RenderedMachineOSImage MachineOSImageReference `json:"renderedMachineOSImage"`
+	// startTime describes when this build began
+	// +kubebuilder:validation:Required
+	StartTime *metav1.Time `json:"startTime"`
+	// endTime describes when the build ended
+	// +optional
+	EndTime *metav1.Time `json:"endTime"`
+	// buildHistory contains previous iterations of failed or interrupted builds related to this one.
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	BuildHistory []BuildHistory `json:"buildHistory,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
+}
+
+// BuildProgess highlights some of the key phases of a build to be tracked in Conditions.
+type BuildProgress string
+
+const (
+	// prepared indicates that the build has finished preparing. A build is prepared
+	// by gathering the build inputs, validating them, and making sure we can do an update as specified.
+	MachineOSBuildPrepared BuildProgress = "Prepared"
+	// building indicates that the build has been kicked off with the specified image builder
+	MachineOSBuilding BuildProgress = "Building"
+	// failed indicates that during the build or preparation process, the build failed.
+	MachineOSBuildFailed BuildProgress = "Failed"
+	// interrupted indicates that the user stopped the build process by modifying part of the build config
+	MachineOSBuildInterrupted BuildProgress = "Interrupted"
```

Comment:
I know there's a ton of comments here and sorry if this is duplicating existing feedback but at a high level a goal I think we should have here is that things related to the booted host container should "feel the same" as "app" containers - where it makes sense.

Most of the other states here apply equally well to application container images.  So why not just import and use e.g. https://github.com/openshift/api/blob/master/build/v1/consts.go#L153 ?

Reply:
@cgwalters sure, I'd be fine with using some of those. I am not sure if there is a 1:1 match for some of the states (like InProgress, Restarted). But Interrupted could map well to Cancelled and Failed to the one you linked above. 

Would you be ok with picking and choosing the ones that match with some new custom states here?

Reply:
We want users to be able to track some sort of "progress". Since we are giving them the ability to control their build inputs but we actually run the build, we don't want to create too much of a black box that leads to more MCO bugs if that makes sense? So states like `Building` and `BuildRestarted`, along with the rationale of what happened specific to MCO, is important

Reply:
Sorry to also comment inline here but seems possibly appropriate.

If the Controller reported a Degraded true status on build failure with a reasonable message. Couldn't debug be done via container logs and or assisted by events? That seems like a reasonable/existing pattern. Status isn't really a full logger just a high-level signal of where to look next? Then also does the user need information if things are good or only when they are not?

Reply:
yeah, it is reasonable to assume they will debug into the container/pod.

Again, I do think that the user would benefit from this flow though given the fact that they have no idea what is going on in the build. 

With our MCP object (and a few others) we tend to update state in both good and bad scenarios, so I'd like to keep this object operating similarly if possible. 

I believe we can change these state constants in tech preview. So should we continue as is on the state reporting front and change as needed in future iterations so we don't get blocked on this?

Reply:
We are opting to use these types for a few reasons

1. possible deprecation of the build API in the future ( we don't want to be dependent on those deprecated types)
2. Using custom types as opposed to re-using other people's types is better practice in the API where applicable.
3. not all of the types seem to fit for what we need.

Open to discussing this more, but for purposes of getting this into 4.16, this is what the team has discussed.

cc @mrunalp @hexfusion @cheesesashimi